From 6f10a1c45d5c29f9480d664c60ac924a8ab94006 Mon Sep 17 00:00:00 2001
From: Thomas Wu <quic_wthomas@quicinc.com>
Date: Tue, 21 Nov 2023 11:53:41 -0800
Subject: [PATCH] wifi: ath12k: Process HT Action Frames in 6 GHz

Management frames for 6 GHz do not include HT Capabilities. This
results in HT Action Frames being dropped in ieee80211_rx_h_action()
because current condition only checks for ht_cap.ht_supported.
Enable processing for these action frames when he_cap.has_he is also true.
This is required to support IEEE Std 802.11ax-2021 26.14.4 "HE dynamic SM power save".

Signed-off-by: Thomas Wu <quic_wthomas@quicinc.com>
---
 net/mac80211/rx.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index af31657..d1931b5 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -3684,8 +3684,9 @@ ieee80211_rx_h_action(struct ieee80211_rx_data *rx)
 
 	switch (mgmt->u.action.category) {
 	case WLAN_CATEGORY_HT:
-		/* reject HT action frames from stations not supporting HT */
-		if (!rx->link_sta->pub->ht_cap.ht_supported)
+		/* reject HT action frames from stations not supporting HT or not HE Capable */
+		if (!rx->link_sta->pub->ht_cap.ht_supported &&
+		    !rx->link_sta->pub->he_cap.has_he)
 			goto invalid;
 
 		if (sdata->vif.type != NL80211_IFTYPE_STATION &&
-- 
2.25.1

